File: backend/app/core/validator.py
from typing import List
from .spec import AgentSkillSpec
from .platforms import is_supported


class ValidationError(Exception):
    def __init__(self, errors: List[str]):
        self.errors = errors
        super().__init__("; ".join(errors))


def validate_spec(spec: AgentSkillSpec) -> None:
    errors: List[str] = []

    if not spec.name:
        errors.append("name is required")
    if not spec.description:
        errors.append("description is required")
    if not spec.triggers.keywords and not spec.triggers.intent:
        errors.append("triggers must contain at least one keyword or intent")
    if not spec.workflow:
        errors.append("workflow must contain at least one step")
    if not spec.constraints:
        errors.append("constraints must not be empty")

    if not is_supported(spec.source_platform):
        errors.append(f"unsupported source platform: {spec.source_platform}")
    if spec.target_platform and not is_supported(spec.target_platform):
        errors.append(f"unsupported target platform: {spec.target_platform}")

    if errors:
        raise ValidationError(errors)

File: frontend/src/lib/validator.ts
import type { AgentSkillSpec, LossReport } from './types';

export class ValidationError extends Error {
  errors: string[];
  constructor(errors: string[]) {
    super(errors.join('; '));
    this.errors = errors;
  }
}

export function validateSpec(spec: AgentSkillSpec): void {
  const errors: string[] = [];
  if (!spec.name) errors.push('缺少 name 字段');
  if (!spec.description) errors.push('缺少 description 字段');
  // If there are no sections at all, and no triggers, that's an error
  if (!spec.triggers.keywords.length && !spec.triggers.intent && !spec.sections.length) {
    errors.push('缺少触发条件（triggers）和章节内容，至少需要一项');
  }
  if (!spec.workflow.length && !spec.sections.length) {
    errors.push('缺少工作流（workflow）或章节（sections）内容');
  }
  if (errors.length) throw new ValidationError(errors);
}

export function buildLossReport(preserved: string[], partial: string[], lost: string[], manualCheck: string[]): LossReport {
  return { preserved, partial, lost, manual_check: manualCheck };
}